In [1]:
!pip install biopython

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 12.4 MB/s eta 0:00:00


In [2]:
from Bio import pairwise2
from Bio.pairwise2 import format_alignment

/usr/local/lib/python3.10/dist-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.
  warnings.warn(


In [3]:
cox1_s_cerevisiae = """
MVQRWLYSTNAKDIAVLYFMLAIFSGMAGTAMSLIIRLELAAPGSQYLHGNSQLFNVLVVGHAVLMIFFL
VMPALIGGFGNYLLPLMIGATDTAFPRINNIAFWVLPMGLVCLVTSTLVESGAGTGWTVYPPLSSIQAHS
GPSVDLAIFALHLTSISSLLGAINFIVTTLNMRTNGMTMHKLPLFVWSIFITAFLLLLSLPVLSAGITML
LLDRNFNTSFFEVSGGGDPILYEHLFWFFGHPEVYILIIPGFGIISHVVSTYSKKPVFGEISMVYAMASI
GLLGFLVWSHHMYIVGLDADTRAYFTSATMIIAIPTGIKIFSWLATIHGGSIRLATPMLYAIAFLFLFTM
GGLTGVALANASLDVAFHDTYYVVGHFHYVLSMGAIFSLFAGYYYWSPQILGLNYNEKLAQIQFWLIFIG
ANVIFFPMHFLGINGMPRRIPDYPDAFAGWNYVASIGSFIATLSLFLFIYILYDQLVNGLNNKVNNKSVI
YNKAPDFVESNTIFNLNTVKSSSIEFLLTSPPAVHSFNTPAVQS
"""
cox1_c_albicans = """
MSYVTRWLFSTSHKDIGILYLIYGMVSAMVATGMSVIIRLELSGPGPMFLHGNNQVFNVLVTGHAIAMIF
LFVMPILIGSFGNYFLPIMIGAVDMAFARLNNISFWCLPPALVCVIASVLIETGAGTGWTVYPPLSSISA
HSGPSVDLAIFAIHLTSISSLLGAINFIVTTLNMRSIGIHMIDMPLFVWAIFFTAILLLLSLPVLTAGVT
LLLMDRNFNTGFYEVAAGGDPILYEHLFYFFGHPEVYIIIIPGFGIISHVISTYTKKPIFGQIGMIYAIG
SIGLLGFLVWSHHMYVVGLDIDSRAYFTSATMVIAIPTGIKIFSWLATLYGGELRLAVPMLFALGFLFLF
TIGGLTGVMLSNASIDVAFHDTYYVVGHFHYVLSMGALFSLIAGYYYWGPAMFGLKYNKVLAEVHYWLLF
VSVNIIFLPMHFLGLNGMPRRIPQYPDAFVGWNVVSSWGSIMSVISVLIGLYSVLVQLTNGENEREEIQV
TPDYLESNNTRDVRDSDLELITSRPAQYHTFSELPILIQQI
"""

First, the two sequences will be aligned globally, with default parameters. There are no gap penalties and identical characters are scored 1, else 0.

In [7]:
global_alignments = pairwise2.align.globalxx(cox1_s_cerevisiae, cox1_c_albicans)
print(format_alignment(*global_alignments[0]))


M--VQ-RWLY-STNA--KDIAV--LYFMLAIFS-GM--AGT--A--MSL-IIRLELAA--PGSQY---LHGNS-QL-FNVLVV-GHAVL--MIFFL
--VMPA-LIGG-FGNYL-LPL-MIGAT-DT-AFP-RI-NNIA-FWV-LPMG--LVCLVT--STLV--ES-GAGTGWTVYPPLSSIQ-A-HS
GPSVDLAIFAL-HLTSISSLLGAINFIVTTLNMRTN--G--MT--MHKLPLFVWS-IFI-TAF-LLLLSLPVLS-AGI-TML
LL--DRNFNTSF-F-EVSG--GGDPILYEHLFW-FFGHPEVYILII-PGFGIISHVV-STYS-KKPV-FGE-IS-MV-YAMASI-
--GLLGFLVWSHHMYIV-GLDA-DT-RAYFTSATMI-IAIPTGIKIFSWLATIH--GGSI--RLAT-PMLYAI-A--FLFLFTM
--GGLTGVA-LA-NASL-DVAFHDTYYVVGHFHYVLSMGAI-FSLF-AGYYYWS-PQIL---GLN-YNEK-LAQIQF----WLI-FIG
ANV---IF-F-PMHFLGI-NGMPRRIPD-YPDAFA-GWNYVA-SI--GSFIATL-SLFLF-IY---I-LYDQL-VNGLNNKV---N--NKSV----I--
YNKA-PDFV--ESNTIFNLNT--VKSS--SIEF-L-L-TSP-PAV--HSFNT----PAV---QS--

||  |  |||  ||    |||    ||  | |   ||  |    |  ||  ||||||    ||      ||||  |  |||| | |||    || | |  |||  || | ||||  ||  ||||  |  ||  |  |||  ||  ||    ||| |   |  |  |  |||||||||||||||  | || ||||||||||  ||||||||||||||||||||||    |  |   |   |||||  ||  ||  |||||||||  ||  |  |||  ||||||   | ||    |||||

In [13]:
# This code finds and printsthe best global alignment between the given sequences.
# Identical characters are given a score of 1. No points are deducted for mismatches and gaps.

for a in global_alignments:
  print(format_alignment(*a))


M--VQ-RWLY-STNA--KDIAV--LYFMLAIFS-GM--AGT--A--MSL-IIRLELAA--PGSQY---LHGNS-QL-FNVLVV-GHAVL--MIFFL
--VMPA-LIGG-FGNYL-LPL-MIGAT-DT-AFP-RI-NNIA-FWV-LPMG--LVCLVT--STLV--ES-GAGTGWTVYPPLSSIQ-A-HS
GPSVDLAIFAL-HLTSISSLLGAINFIVTTLNMRTN--G--MT--MHKLPLFVWS-IFI-TAF-LLLLSLPVLS-AGI-TML
LL--DRNFNTSF-F-EVSG--GGDPILYEHLFW-FFGHPEVYILII-PGFGIISHVV-STYS-KKPV-FGE-IS-MV-YAMASI-
--GLLGFLVWSHHMYIV-GLDA-DT-RAYFTSATMI-IAIPTGIKIFSWLATIH--GGSI--RLAT-PMLYAI-A--FLFLFTM
--GGLTGVA-LA-NASL-DVAFHDTYYVVGHFHYVLSMGAI-FSLF-AGYYYWS-PQIL---GLN-YNEK-LAQIQF----WLI-FIG
ANV---IF-F-PMHFLGI-NGMPRRIPD-YPDAFA-GWNYVA-SI--GSFIATL-SLFLF-IY---I-LYDQL-VNGLNNKV---N--NKSV----I--
YNKA-PDFV--ESNTIFNLNT--VKSS--SIEF-L-L-TSP-PAV--HSFNT----PAV---QS--

||  |  |||  ||    |||    ||  | |   ||  |    |  ||  ||||||    ||      ||||  |  |||| | |||    || | |  |||  || | ||||  ||  ||||  |  ||  |  |||  ||  ||    ||| |   |  |  |  |||||||||||||||  | || ||||||||||  ||||||||||||||||||||||    |  |   |   |||||  ||  ||  |||||||||  ||  |  |||  ||||||   | ||    |||||

Next, they will be aligned locally, with default parameters. There are no gap penalties and identical characters are scored 1, else 0.

In [9]:
local_alignments = pairwise2.align.localxx(cox1_s_cerevisiae, cox1_c_albicans)

In [14]:
# This code finds and printsthe best global alignment between the given sequences.
# Identical characters are given a score of 1. No points are deducted for mismatches and gaps.


for a in local_alignments:
  print(format_alignment(*a))


M--VQ-RWLY-STNA--KDIAV--LYFMLAIFS-GM--AGT--A--MSL-IIRLELAA--PGSQY---LHGNS-QL-FNVLVV-GHAVL--MIFFL
--VMPA-LIGG-FGNYL-LPL-MIGAT-DT-AFP-RI-NNIA-FWV-LPMG--LVCLVT--STLV--ES-GAGTGWTVYPPLSSIQ-A-HS
GPSVDLAIFAL-HLTSISSLLGAINFIVTTLNMRTN--G--MT--MHKLPLFVWS-IFI-TAF-LLLLSLPVLS-AGI-TML
LL--DRNFNTSF-F-EVSG--GGDPILYEHLFW-FFGHPEVYILII-PGFGIISHVV-STYS-KKPV-FGE-IS-MV-YAMASI-
--GLLGFLVWSHHMYIV-GLDA-DT-RAYFTSATMI-IAIPTGIKIFSWLATIH--GGSI--RLAT-PMLYAI-A--FLFLFTM
--GGLTGVA-LA-NASL-DVAFHDTYYVVGHFHYVLSMGAI-FSLF-AGYYYWS-PQIL---GLN-YNEK-LAQIQF----WLI-FIG
ANV---IF-F-PMHFLGI-NGMPRRIPD-YPDAFA-GWNYVA-SI--GSFIATL-SLFLF-IY---I-LYDQL-VNGLNNKV---N--NKSV----I--
YNKA-PDFV--ESNTIFNLNT--VKSS--SIEF-L-L-TSP-PAV--HSFNT----PAV---QS--

||  |  |||  ||    |||    ||  | |   ||  |    |  ||  ||||||    ||      ||||  |  |||| | |||    || | |  |||  || | ||||  ||  ||||  |  ||  |  |||  ||  ||    ||| |   |  |  |  |||||||||||||||  | || ||||||||||  ||||||||||||||||||||||    |  |   |   |||||  ||  ||  |||||||||  ||  |  |||  ||||||   | ||    |||||

In [16]:
print(len(cox1_s_cerevisiae))
print(len(cox1_c_albicans))

543
540


Identical matches between the two sequences were given a score of 1. All other gaps and mismatches were not penalised and were merely scored 0. The score from the current alignment.

Both the local and global alignments produced a score of 383. This means that there are 383 amino acid residues that are identical between the two amino acid sequences.

In [17]:
total_residues = len(cox1_s_cerevisiae) + len(cox1_c_albicans)
identical_prop = 383 / total_residues
print(identical_prop)

0.3536472760849492


Approximately 35% of the total compared amino acids are identical.

The two sequences were sourced from the [NCBI](https://www.ncbi.nlm.nih.gov/).

The documentation for the packages used was obtained from the [Biopython website](https://biopython.org/docs/1.76/api/Bio.Align.html).